File: cpp/examples/linux/accelerometer/ADXL345/demo/main.cpp
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <unistd.h>
#include "I2CConnectionLinux.h"
#include "ADXL345.h"

int main() {
    const char* bus_env  = getenv("I2C_BUS");
    const char* addr_env = getenv("I2C_ADDR");
    int     bus  = bus_env  ? atoi(bus_env)       : 1;
    uint8_t addr = addr_env ? (uint8_t)strtol(addr_env, nullptr, 0) : 0x53;
    I2CConnectionLinux connection(bus, addr);

    ADXL345Minimal accel(connection);                       // Create ADXL345 driver, (connection, spi=false)

    // --- 50-sample stationary tilt characterization at 10 Hz ---
    // With the sensor flat and the Z axis up, gravity should project entirely
    // onto Z. Tilting the board visibly redistributes the 1 *g* magnitude
    // across X and Y; the total vector magnitude stays near 1 *g*.
    const int SAMPLES = 50;

    float mag_min = 1e9f, mag_max = -1e9f;

    for (int n = 0; n < SAMPLES; n++) {
        float x, y, z;
        accel.read(x, y, z);                                // Read 3-axis acceleration, (x, y, z) → g, g, g
        float mag = sqrtf(x * x + y * y + z * z);
        if (mag < mag_min) mag_min = mag;
        if (mag > mag_max) mag_max = mag;
        printf("%2d  x=%+.3f  y=%+.3f  z=%+.3f  |a|=%.3f g\n",
               n, (double)x, (double)y, (double)z, (double)mag);
        usleep(100000);
    }

    printf("min |a|=%.3f g  max |a|=%.3f g\n", (double)mag_min, (double)mag_max);
    return 0;
}

File: rust/examples/linux/accelerometer/adxl345/demo/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::accelerometer::Adxl345Minimal;

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x53);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut accel = Adxl345Minimal::new(dev, addr, false).expect("init ADXL345"); // Create ADXL345 driver, (i2c, addr=0x53, spi=false)

    // --- 50-sample stationary tilt characterization at 10 Hz ---
    // With the sensor flat and the Z axis up, gravity should project entirely
    // onto Z. Tilting the board visibly redistributes the 1 *g* magnitude
    // across X and Y; the total vector magnitude stays near 1 *g*.
    const SAMPLES: usize = 50;
    const PERIOD_MS: u64 = 100;

    let mut mag_min = f32::INFINITY;
    let mut mag_max = f32::NEG_INFINITY;

    for n in 0..SAMPLES {
        let (x, y, z) = accel.read().expect("read acceleration");   // Read 3-axis acceleration, () → (f32, f32, f32) g
        let mag = (x * x + y * y + z * z).sqrt();
        if mag < mag_min { mag_min = mag; }
        if mag > mag_max { mag_max = mag; }
        println!("{:2}  x={:+.3}  y={:+.3}  z={:+.3}  |a|={:.3} g",
                 n, x, y, z, mag);
        std::thread::sleep(std::time::Duration::from_millis(PERIOD_MS));
    }

    println!("min |a|={:.3} g  max |a|={:.3} g", mag_min, mag_max);
}